<a href="https://colab.research.google.com/github/Katonokatono/Suicide/blob/Preprocessing/Undersample_%26_Oversampling_Suicide_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np


In [2]:
df=pd.read_csv('/content/final_data_suicide.csv')
df[:5]

Unnamed: 0  ...                                         clean_text
0           1  ...  bunmo multi item stretchy string fidget toy pk...
1           2  ...  unkonfined basic motivation behind behavior bi...
2           3  ...  munenekimathi zablonorina bp normal yes check ...
3           4  ...  deal anxious share experience anxiety yvonne m...
4           5  ...  know dry spell cause anxiety depression even s...

[5 rows x 9 columns]

In [3]:
df['label'].value_counts()

general tweet      2610
awareness           478
anxiety             357
depression          355
substance abuse     134
thoughts            129
stress              109
Name: label, dtype: int64

In [4]:
# Create label dataframes
df_gen_tweet=df[df['label']=='general tweet']
df_depression=df[df['label']=='depression']
df_anxiety=df[df['label']=='anxiety']
df_stress=df[df['label']=='stress']
df_awareness=df[df['label']=='awareness']
df_substance_abuse=df[df['label']=='substance abuse']
df_thoughts=df[df['label']=='thoughts']

In [21]:
# Class counts
count_class_0, count_class_1,count_class_2,count_class_3,count_class_4,count_class_5,count_class_6 = df.label.value_counts()

# View class counts
count_class_0, count_class_1,count_class_2,count_class_3,count_class_4,count_class_5,count_class_6

(2610, 478, 357, 355, 134, 129, 109)

We use depression class as it is our midpoint with 355

In [11]:
count_class_3

355

undersampling

In [22]:
# Undersample class and concat the DataFrames of both class
df_gen_tweet_under = df_gen_tweet.sample(count_class_3)
df_awareness_under = df_awareness.sample(count_class_3)
df_anxiety_under = df_anxiety.sample(count_class_3)
df_under = pd.concat([df_gen_tweet_under,df_awareness_under], axis=0)

print('Random under-sampling:')
print(df_under.label.value_counts())

Random under-sampling:
awareness        355
general tweet    355
Name: label, dtype: int64


Oversampling

In [15]:
# Ondersample 0-class and concat the DataFrames of both class
df_gen_tweet_under = df_gen_tweet.sample(count_class_3)
df_awareness_under = df_awareness.sample(count_class_3)
df_anxiety_under = df_anxiety.sample(count_class_3)
df_under = pd.concat([df_gen_tweet_under,df_anxiety_under,df_awareness_under], axis=0)

print('Random under-sampling:')
print(df_under.label.value_counts())

Random under-sampling:
anxiety          355
awareness        355
general tweet    355
Name: label, dtype: int64


Oversampling

In [16]:
# Ondersample 0-class and concat the DataFrames of both class
df_thoughts_over= df_thoughts.sample(count_class_3, replace=True)
df_substance_over= df_substance_abuse.sample(count_class_3, replace=True)
df_stress_over= df_stress.sample(count_class_3, replace=True)

df_over = pd.concat([df_thoughts_over,df_substance_over,df_stress_over], axis=0)



print('Random under-sampling:')
print(df_over.label.value_counts())

Random under-sampling:
substance abuse    355
stress             355
thoughts           355
Name: label, dtype: int64


In [20]:
df_balanced = pd.concat([df_over,df_under,df_depression], axis=0)
df_balanced['label'].value_counts()

depression         355
thoughts           355
awareness          355
substance abuse    355
anxiety            355
stress             355
general tweet      355
Name: label, dtype: int64

In [24]:
df_balanced.to_csv('balanced_data')